probert/filesystem.py

```diff
@@ -0,0 +1,64 @@
+# Copyright 2019 Canonical, Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+import os
+import pyudev
+
+log = logging.getLogger('probert.filesystems')
+
+
+def get_supported_filesystems():
+    """ Return a list of filesystems that the kernel currently supports
+        as read from /proc/filesystems.
+
+        Raises RuntimeError if /proc/filesystems does not exist.
+    """
+    proc_fs = "/proc/filesystems"
+    if not os.path.exists(proc_fs):
+        raise RuntimeError("Unable to read 'filesystems' from %s" % proc_fs)
+
+    with open(proc_fs, 'r') as fh:
+        proc_fs_contents = fh.read()
+
+    return [l.split('\t')[1].strip()
+            for l in proc_fs_contents.splitlines()]
+
+
+def get_device_filesystem(device):
+    # extract ID_FS_* keys into dict, dropping leading ID_FS
+    return {k.replace('ID_FS_', ''): v
+            for k, v in device.items() if k.startswith('ID_FS_')}
+
+
+def probe(context=None):
+    """ Capture detected filesystems found on discovered block devices.
+        Filter the detected values via the host kernel's supported file
+        systems.
+    """
+    filesystems = {}
+    supported_fs = get_supported_filesystems()
+    if not context:
+        context = pyudev.Context()
+
+    for device in context.list_devices(subsystem='block'):
+        # Ignore block major=1 (ramdisk) and major=7 (loopback)
+        # these won't ever be used in recreating storage on target systems.
+        if device['MAJOR'] not in ["1", "7"]:
+            fs_info = get_device_filesystem(device)
+            if fs_info and fs_info['TYPE'] in supported_fs:
```

Comment:
'swap' does not appear to be in /proc/filesystems but can show up as ID_FS_TYPE, I guess we should hard code that one.

This code didn't detect an ntfs external drive I had plugged in -- it shows up as fuseblk in the output of mount but somehow it doesn't get a ID_FS_TYPE in udev. That's perhaps OK but because probert did find the mount, the probert to curtin code complained (well the generated config did not validate).

Reply:
Good point on the swap;  I think we can check for /proc/swaps, which should be present on kernels with CONFIG_SWAP enabled.

Fuseblk is new to me; I think it would be good to add that; just need some info from udev and possible sysfs to see what we can pick out to detect how we identify them (via serial or wwn or something else, path I suppose).

Reply:
It's hard to imagine an Ubuntu kernel without CONFIG_SWAP but sure :)

And it turns out I was wrong about the ntfs drive missing ID_FS_TYPE in udev, it's there OK:

```
(master)mwhudson@ringil:~/src/probert$ udevadm info  /dev/disk/by-id/usb-Seagate_BUP_Slim_BK_NA7X3MG1-0\:0-part1 | grep FS_TYPE
E: ID_FS_TYPE=ntfs
```

(I guess I was looking at the udev data for the drive not the partition).

Given that it turns out that both of these problems would be solved by _not_ filtering the fs found through /proc/filesystems, perhaps we shouldn't do that? I guess this would mean there's a chance that we end up presenting filesystems we can't actually mount but that might be the better of the two alternatives?

Reply:
I think it best to not present things we can't mount.  That said, we can and do mount NTFS, so theres a gap w.r.t how we detect this.  What does your /proc/filesystems show?

Reply:
```
$ cat /proc/filesystems 
nodev	sysfs
nodev	rootfs
nodev	ramfs
nodev	bdev
nodev	proc
nodev	cpuset
nodev	cgroup
nodev	cgroup2
nodev	tmpfs
nodev	devtmpfs
nodev	configfs
nodev	debugfs
nodev	tracefs
nodev	securityfs
nodev	sockfs
nodev	dax
nodev	bpf
nodev	pipefs
nodev	hugetlbfs
nodev	devpts
	ext3
	ext2
	ext4
	squashfs
	vfat
nodev	ecryptfs
	fuseblk
nodev	fuse
nodev	fusectl
nodev	pstore
nodev	efivarfs
nodev	mqueue
	btrfs
nodev	autofs
nodev	zfs
nodev	binfmt_misc
nodev	overlay
	iso9660
```

I guess there must be some way or enumerating or at least probing for which fuse filesystems are available but I don't know what it might be... time passes... well, basically it seems udisks2 hardcodes a list https://github.com/storaged-project/udisks/blob/udisks-2.8.2/src/udiskslinuxfilesystem.c#L224-L243 which is, well, lovely. I guess we could/should copy that list?

Reply:
Having said that, I'm not sure that's what that list is for, I think it might be more of a whitelist to valid arguments againts. I think ultimately the mounting of ntfs (resp exfat) happens because I have a mount.ntfs (resp mount.exfat) binary. So "$fs_type in /proc/filesystems or "/sbin/mount.$fs_type exists" ? And still a special case for swap.

Reply:
Why isn't  the block device  path for your usb disk in pyudev context?

Can you provide this?
udevadm info  /dev/disk/by-id/usb-Seagate_BUP_Slim_BK_NA7X3MG1-0

The filtering was meant to ignore filesystems that we cannot create.  In the context of the probing output I think we could drop the filtering.   The only storage-config element where this matters is in type:format where we create filesystems;  generally type:mount doesn't nor should care since we support specifying mounts via spec.

Swap still needs special case; and /proc/swaps isn't quite right; I have code which checks for swap signature on blockdevices which is more useful since it is possible to boot a kernel without swap but the target systems can use swap.